File: src/portfolio/rst_utils.py
from io import TextIOWrapper


def title(f: TextIOWrapper, t: str):
    f.write("=" * len(t) + "\n")
    f.write(t + "\n")
    f.write("=" * len(t) + "\n\n")


def subtitle(f: TextIOWrapper, t: str):
    f.write(t + "\n")
    f.write("=" * len(t) + "\n\n")


def indexes_and_tables(f: TextIOWrapper):
    f.write("* :ref:`genindex`\n\n")


def card(
    f: TextIOWrapper,
    title: str | None = None,
    body: str | None = None,
    link: str | None = None,
    indent: int = 0,
):
    indent_str = "\t" * indent
    f.write(indent_str + ".. card::")
    if title is not None:
        f.write(" " + title)
    if body is None and link is None:
        f.write("\n\n")
    else:
        if link is not None:
            f.write(f"\n\t{indent_str}:link: {link}")

            if body is None:
                f.write("\n\n")

        if body is not None:
            body = body.replace("\n", indent_str + "\n\n\t")
            f.write(f"\n\n\t{indent_str}{body}\n\n")


def card_carousel(
    f: TextIOWrapper,
    titles: list,
    bodies: list,
    links: list,
    num_show: int = 5,
):
    f.write(f".. card-carousel:: {num_show}\n\n")
    for title, body, link in zip(titles, bodies, links):
        card(f, title, body, link, indent=1)

File: portfolio/src/website/rst_utils.rs
pub trait RstUtils {
    fn add_title(&mut self, title: String) -> &mut Self;
    fn add_subtitle(&mut self, subtitle: String) -> &mut Self;
    fn add_card(
        &mut self,
        title: Option<String>,
        body: Option<String>,
        link: Option<String>,
        indent: i32,
    ) -> &mut Self;
    fn add_card_carousel(
        &mut self,
        titles: Vec<Option<String>>,
        bodies: Vec<Option<String>>,
        links: Vec<Option<String>>,
        num_show: i32,
    ) -> &mut Self;
}

impl RstUtils for String {
    fn add_title(&mut self, title: String) -> &mut Self {
        let bars = "=".repeat(title.len());
        self.push_str(format!("{bars}\n{title}\n{bars}\n\n").as_str());
        self
    }

    fn add_subtitle(&mut self, subtitle: String) -> &mut Self {
        let bars = "=".repeat(subtitle.len());
        self.push_str(format!("{subtitle}\n{bars}\n\n").as_str());
        self
    }

    fn add_card(
        &mut self,
        title: Option<String>,
        body: Option<String>,
        link: Option<String>,
        indent: i32,
    ) -> &mut Self {
        let indent_str = "\t".repeat(indent as usize);
        self.push_str(format!("{indent_str}.. card::").as_str());
        if title.is_some() {
            self.push_str(format!(" {}", title.unwrap()).as_str());
        }
        if body.is_none() && link.is_none() {
            self.push_str("\n\n");
        } else {
            if link.is_some() {
                self.push_str(format!("\n\t{indent_str}:link: {}", link.unwrap()).as_str());

                if body.is_none() {
                    self.push_str("\n\n");
                }
            }

            if body.is_some() {
                let body2 = body
                    .unwrap()
                    .replace("\n", format!("{indent_str}\n\n\t").as_str());
                self.push_str(format!("\n\n\t{indent_str}{body2}\n\n").as_str())
            }
        }
        self
    }

    fn add_card_carousel(
        &mut self,
        titles: Vec<Option<String>>,
        bodies: Vec<Option<String>>,
        links: Vec<Option<String>>,
        num_show: i32,
    ) -> &mut String {
        self.push_str(format!(".. card-carousel:: {num_show}\n\n").as_str());
        for ((title, body), link) in titles.into_iter().zip(bodies).zip(links) {
            self.add_card(title, body, link, 1);
        }
        self
    }
}
